simplex/block_builder_test.go

```diff
@@ -140,3 +140,39 @@ func TestIncomingBlock(t *testing.T) {
 	bb.IncomingBlock(ctx)
 	require.Equal(t, 1, count)
 }
+
+func TestBlockBuildingExponentialBackoff(t *testing.T) {
+	ctx := context.Background()
+	vm := newTestVM()
+	genesis := newTestBlock(t, newBlockConfig{})
+	child := newTestBlock(t, newBlockConfig{
+		prev: genesis,
+	})
+
+	count := 0
+	vm.WaitForEventF = func(_ context.Context) (common.Message, error) {
+		count++
+		return common.PendingTxs, nil
+	}
+
+	vm.BuildBlockF = func(_ context.Context) (snowman.Block, error) {
+		return nil, errors.New("failed to build block")
+	}
+
+	bb := &BlockBuilder{
+		log:          logging.NoLog{},
+		vm:           vm,
+		blockTracker: genesis.blockTracker,
+	}
+
+	timeoutCtx, cancelCtx := context.WithTimeout(ctx, time.Second)
+	defer cancelCtx()
+
+	block, built := bb.BuildBlock(timeoutCtx, child.BlockHeader().ProtocolMetadata)
+	require.False(t, built)
+	require.Nil(t, block)
+
+	// expected number of count increases
+	// 10, 20, 40, 80, 160, 320, 640 = 7 attempts
```

Comment:
I guess these attempts are due to the timeout? In that case, I think it's better to just control the amount of retries in the BuildBlockF itself (based on the count), otherwise if the CI runner is overloaded we might have less retries because of lag.... thoughts?